### DISCUSSION TITLE
Benchmark framework

### DISCUSSION THREAD

**Discussion by praksharma** [praksharma/DeepINN#46] [Ideas]
DeepINN started as a lightweight framework for developing Physics-Informed Neural Networks (PINNs).  Over the last few years I have realised the PINN ecosystem suffers from several issues:

* Every paper introduces a new implementation.
* Benchmarking is difficult because every paper defines its own training protocol.
* Existing benchmark repositories are largely paper-specific and are not designed to be extended by third parties.
* Many promising research repositories become unmaintained after publication.

PINNacle is a great example: https://pinnacle-docs.vercel.app/

* The github repo is unorganised.
* You can't easily add a new architecture coz they use DeepXDE as backend where your architecture should be `dde.nn.FNN` instance. So, unless you understand how python modules work you can't simply add a framework.
* Adding external geometry is not easy. It mostly relies on CSG geometry module of DeepXDE. DeepXDE does allow anchor points but the is not very obvious unless you dive into the DeepXDE docs.
* Loss function modification isn't easy as well. So things like self-adaptive loss coefficient is a no-go for non-coders. For example, DeeXDE doesn't allow you inputting `nn.Parameters()` in the loss function unless you modify the source code.

Another example is PDEBench. But it suffers the same problem, that it is based on DeepXDE.

I want something similar to computer vision where they compare the model performance nicely. DeepINN should provide:

* Reference implementations of common PINN methods
* Standard benchmark problems
* Reproducible training pipelines
* Extensible plugin architecture
* Benchmark reports
* Framework-independent APIs wherever possible
* A weights and bias integration toggle
 
For example, if I want to benchmark a Poisson equation, I should only implement my method.

```
problem = registry.problem("poisson2d_classic")
```
And then simply run the benchmark.

```
pinnbench run --problem poisson2d --method my_method --seed 42
```

The benchmark should already know:

* the PDE,
* the geometry,
* the boundary conditions,
* the sampling strategy,
* the evaluation protocol,
* and the metrics.

The user should only provide their implementation of the PINN method, not redefine the benchmark itself.
